### ISSUE TITLE
type error when specifiying a connectionTimeout for HttpClient

### ISSUE THREAD

**Issue by miDeb** [dart-lang/sdk#44895]
Minimal code example:
```dart
import 'dart:io';

void main() async {
  final client = HttpClient();
  client.connectionTimeout = Duration();
  await client.openUrl(
    'get',
    Uri.parse(
      'https://example.com/',
    ),
  );
}

```

results in:
```
Unhandled exception:
type '() => Null' is not a subtype of type '(() => FutureOr<SecureSocket>)?' of 'onTimeout'
#0      Future.timeout (dart:async/future_impl.dart)
#1      _ConnectionTarget.connect.<anonymous closure> (dart:_http/http_impl.dart:2222:37)
<asynchronous suspension>
#2      _ConnectionTarget.connect.<anonymous closure> (dart:_http/http_impl.dart)
<asynchronous suspension>
```
OS: linux
dart: `Dart SDK version: 2.13.0-0.0.dev (dev) (Thu Feb 4 08:08:19 2021 -0800) on "linux_x64"`

**Comment by mit-mit** [comment]
cc @lrhn

**Comment by miDeb** [comment]
I believe the problem [is the following:](https://github.com/dart-lang/sdk/blob/master/sdk/lib/_http/http_impl.dart#L2219)
```dart
Future socketFuture = task.socket; // type of task.socket: Future<Socket>
// ...
// but socketFuture is used as a Future<Socket?>, if there is a timeout it returns null:
socketFuture = socketFuture.timeout(connectionTimeout, onTimeout: () {
    _socketTasks.remove(task);
   ask.cancel();
   return null;
});
```

What's the proper way to convert a `Future<T>` to a `Future<T?>` ?

**Comment by lrhn** [comment]
Currently there is no easy way to just introduce nullability in the type of a future.
I'd write this as:
```dart
socketFuture = socketFuture.then<Socket?>((x) => x).timeout(....);
```

We could perhaps introduce a `cast` extension method on futures, so you could just write:
```dat
socketFuture = socketFuture.cast<Socket?>().timeout(...);
```
but we can't restrict it to up-casts, so it's going to be doing dynamic casts of all events, even when not necessary.
The alternative would be a `Future<T?> orNull();` extension method on `Future<T>`.

**Comment by miDeb** [comment]
It is currently unclear when/how a [null safe version of dio](https://github.com/flutterchina/dio/pull/1026) will be released because the mantainer is inactive, but this bug will likely block a release. Is it possible to uplift this fix (if that is necessary)?

**Comment by mit-mit** [comment]
This issue was cherry picked in https://github.com/dart-lang/sdk/issues/44989, and is available in Dart's beta channel in version 2.12.0-259.14.beta and later.
